File: pbd.cpp
#include "pbd.h"

#include <cassert>
#include <fstream>
#include <stdexcept>
#include "pbd_helper.h"

pbd::l_table pbd::sim_to_l_table(
  const parameters& pbd_parameters
)
{
  const std::string r_filename{"run_pbd_sim.R"};
  const std::string csv_filename{"run_pbd_sim.csv"};
  //Create script to create the phylogenies
  {
    std::ofstream f(r_filename);
    f << "library(PBD)\n"
      << "set.seed("<< pbd_parameters.get_seed() << ")\n"
      << "output <- PBD::pbd_sim(\n"
      << "  pars = c(\n"
      << "    " << pbd_parameters.get_birth_good() << ",\n"
      << "    " << pbd_parameters.get_completion() << ",\n"
      << "    " << pbd_parameters.get_birth_incipient() << ",\n"
      << "    " << pbd_parameters.get_death_good() << ",\n"
      << "    " << pbd_parameters.get_death_incipient() << "\n"
      << "  ),\n"
      << "  age = " << pbd_parameters.get_time() << ",\n"
      << "  soc = 2,\n"
      << "  plotit = TRUE\n"
      << ")\n"
      << "dev.off()\n"
      << "write.csv(x = output$L, file = \"" << csv_filename << "\")\n"
    ;
  }
  const int error{
    std::system((std::string("Rscript ") + r_filename).c_str())
  };
  if (error)
  {
    throw std::runtime_error("command failed");
  }
  return load_l_table_from_csv(csv_filename);
}

pbd::l_table pbd::sim_to_l_table(
  const double birth_good,
  const double birth_incipient,
  const double completion,
  const double death_good,
  const double death_incipient,
  const double time,
  const int seed
)
{
  return sim_to_l_table(
    parameters(
      birth_good,
      birth_incipient,
      completion,
      death_good,
      death_incipient,
      time,
      seed
    )
  );
}

pbd::nltt pbd::sim_to_nltt_igtree_extinct(
  const parameters& pbd_parameters
)
{
  const std::string csv_filename{pbd_parameters.get_nltt_plot_filename()};
  sim_to_nltt_igtree_extinct(pbd_parameters, csv_filename);
  return load_nltt_from_csv(csv_filename);
}

void pbd::sim_to_nltt_igtree_extinct(
  const parameters& pbd_parameters,
  const std::string& csv_filename)
{
  //Create script to create the phylogenies
  const std::string r_filename{"sim_to_nltt_igtree_extinct.R"};
  {
    std::ofstream f(r_filename);
    f << "library(PBD)\n"
      << "set.seed("<< pbd_parameters.get_seed() << ")\n"
      << "filename <- \"" << csv_filename << "\"\n"
      << "birth_good <- " << pbd_parameters.get_birth_good() << "\n"
      << "completion <- " << pbd_parameters.get_completion() << "\n"
      << "birth_incipient <- " << pbd_parameters.get_birth_incipient() << "\n"
      << "death_good <- " << pbd_parameters.get_death_good() << "\n"
      << "death_incipient <- " << pbd_parameters.get_death_incipient() << "\n"
      << "age  <- " << pbd_parameters.get_time() << "\n"
      << "out <- PBD::pbd_sim(\n"
      << "  pars = c(\n"
      << "    birth_good,\n"
      << "    completion,\n"
      << "    birth_incipient,\n"
      << "    death_good,\n"
      << "    death_incipient\n"
      << "  ),\n"
      << "  age = age,\n"
      << "  soc = 2\n"
      << ")\n"
      << "\n"
      << "phy <- out$igtree.extinct\n"
      << "xy <- ape::ltt.plot.coords( phy, backward = TRUE, tol = 1e-6)\n"
      << "xy[, 2] <- xy[, 2] / max(xy[, 2])\n"
      << "xy[, 1] <- xy[, 1] + abs( min( xy[, 1]))\n"
      << "xy[, 1] <- xy[, 1] / max( xy[, 1])\n"
      << "\n"
      << "write.csv(x = xy, file = filename)\n"
    ;
  }
  const int error{
    std::system((std::string("Rscript ") + r_filename).c_str())
  };
  if (error)
  {
    throw std::runtime_error("command failed");
  }

}

pbd::nltt pbd::sim_to_nltt_recon(
  const parameters& pbd_parameters
)
{
  const std::string csv_filename{"sim_to_nltt_recon.csv"};
  sim_to_nltt_recon(pbd_parameters, csv_filename);
  return load_nltt_from_csv(csv_filename);
}

void pbd::sim_to_nltt_recon(
  const parameters& pbd_parameters,
  const std::string& csv_filename
)
{
  const std::string r_filename{"only_sim_to_nltt_recon.R"};
  //Create script to create the phylogenies
  {
    std::ofstream f(r_filename);
    f << "library(PBD)\n"
      << "set.seed("<< pbd_parameters.get_seed() << ")\n"
      << "filename <- \"" << csv_filename << "\"\n"
      << "birth_good <- " << pbd_parameters.get_birth_good() << "\n"
      << "completion <- " << pbd_parameters.get_completion() << "\n"
      << "birth_incipient <- " << pbd_parameters.get_birth_incipient() << "\n"
      << "death_good <- " << pbd_parameters.get_death_good() << "\n"
      << "death_incipient <- " << pbd_parameters.get_death_incipient() << "\n"
      << "age  <- " << pbd_parameters.get_time() << "\n"
      << "out <- PBD::pbd_sim(\n"
      << "  pars = c(\n"
      << "    birth_good,\n"
      << "    completion,\n"
      << "    birth_incipient,\n"
      << "    death_good,\n"
      << "    death_incipient\n"
      << "  ),\n"
      << "  age = age,\n"
      << "  soc = 2\n"
      << ")\n"
      << "\n"
      << "phy <- out$recontree\n"
      << "xy <- ape::ltt.plot.coords( phy, backward = TRUE, tol = 1e-6)\n"
      << "xy[, 2] <- xy[, 2] / max(xy[, 2])\n"
      << "xy[, 1] <- xy[, 1] + abs( min( xy[, 1]))\n"
      << "xy[, 1] <- xy[, 1] / max( xy[, 1])\n"
      << "\n"
      << "write.csv(x = xy, file = filename)\n"
    ;
  }
  const int error{
    std::system((std::string("Rscript ") + r_filename).c_str())
  };
  if (error)
  {
    throw std::runtime_error("command failed");
  }
  assert(pbd::is_regular_file(csv_filename));
}

pbd::nltt pbd::sim_to_nltt_recon(
  const double birth_good,
  const double birth_incipient,
  const double completion,
  const double death_good,
  const double death_incipient,
  const double time,
  const int seed
)
{
  return sim_to_nltt_recon(
    parameters(
      birth_good,
      birth_incipient,
      completion,
      death_good,
      death_incipient,
      time,
      seed
    )
  );
}

void pbd::sim_to_png(
  const parameters& pbd_parameters,
  const std::string& png_filename
)
{
  const std::string r_filename{"run_pbd_sim.R"};
  //Create script to create the phylogenies
  {
    std::ofstream f(r_filename);
    f << "library(PBD)\n"
      << "set.seed("<< pbd_parameters.get_seed() << ")\n"
      << "png(\"" << png_filename << "\")\n"
      << "output <- PBD::pbd_sim(\n"
      << "  pars = c(\n"
      << "    " << pbd_parameters.get_birth_good() << ",\n"
      << "    " << pbd_parameters.get_completion() << ",\n"
      << "    " << pbd_parameters.get_birth_incipient() << ",\n"
      << "    " << pbd_parameters.get_death_good() << ",\n"
      << "    " << pbd_parameters.get_death_incipient() << "\n"
      << "  ),\n"
      << "  age = " << pbd_parameters.get_time() << ",\n"
      << "  soc = 2,\n"
      << "  plotit = TRUE\n"
      << ")\n"
      << "dev.off()\n"
    ;
  }
  const int error{
    std::system((std::string("Rscript ") + r_filename).c_str())
  };
  if (error)
  {
    throw std::runtime_error("command failed");
  }
}

void pbd::sim_to_png(
  const double birth_good,
  const double birth_incipient,
  const double completion,
  const double death_good,
  const double death_incipient,
  const double time,
  const int seed,
  const std::string& png_filename
)
{
  return sim_to_png(
    parameters(
      birth_good,
      birth_incipient,
      completion,
      death_good,
      death_incipient,
      time,
      seed
    ),
    png_filename
  );
}

File: pbd.R
# For GUI
create_plot <- function() {

  library(PBD)

  birth_good <- 0.2
  completion <- 0.2
  birth_incipient <- 0.2
  death_good <- 0.1
  death_incipient <- 0.1
  age  <- 15
  filename <- "pbd.png"

  png(filename)
  PBD::pbd_sim(
    pars = c(
      birth_good,
      completion,
      birth_incipient,
      death_good,
      death_incipient
    ),
    age = age,
    soc = 2,
    plotit = TRUE
  )
  dev.off()

}


create_l_table() {
  library(PBD)
  filename <- "pbd.csv"
  birth_good <- 0.2
  completion <- 0.2
  birth_incipient <- 0.2
  death_good <- 0.1
  death_incipient <- 0.1
  age  <- 15

  out <- PBD::pbd_sim(
    pars = c(
      birth_good,
      completion,
      birth_incipient,
      death_good,
      death_incipient
    ),
    age = age,
    soc = 2
  )
  write.csv(x = out$L, file = filename)
}

create_nltt <- function() {

  library(PBD)
  filename <- "nltt_recontree.csv"
  birth_good <- 0.2
  completion <- 0.2
  birth_incipient <- 0.2
  death_good <- 0.1
  death_incipient <- 0.1
  age  <- 15

  out <- PBD::pbd_sim(
    pars = c(
      birth_good,
      completion,
      birth_incipient,
      death_good,
      death_incipient
    ),
    age = age,
    soc = 2
  )

  phy <- out$recontree
  #we use the ltt.plot.coords function from the package ape
  xy <- ape::ltt.plot.coords( phy, backward = TRUE, tol = 1e-6)
  xy[, 2] <- xy[, 2] / max(xy[, 2]) #normalize number lineages

  xy[, 1] <- xy[, 1] + abs( min( xy[, 1])) #make sure time runs from 0..T
  xy[, 1] <- xy[, 1] / max( xy[, 1])      #normalize time

  write.csv(x = xy, file = filename)
}



# calc_nltt_statistic




brts <- branching.times(phylogeny)  # branching times
init_b <- 0.2  # speciation-initiation rate
init_mu_1 <- 0.05  # extinction rate of good species
init_la_1 <- 0.3 # speciation-completion rate
#init_mu_2 <- 0.05  # extinction rate of incipient species  # not used

# The initial values of the parameters that must be optimized
initparsopt <- c(init_b, init_mu_1, init_la_1)

# The extinction rates between incipient and good species are equal
exteq <- TRUE

# The first element of the branching times is the crown age (and not the stem age)
soc <- 2

# Conditioning on non-extinction of the phylogeny
# as I actively selected for a nice phylogeny
cond <- 1

# Give the likelihood of the phylogeny (instead of the likelihood of the branching times)
btorph <- 1
```

Calling the `pbd_ML` function will give a lot of noise. 
It can be turned off by setting verbose = FALSE.
Here goes:

```{r}
r <- pbd_ML(
  brts = brts,
  initparsopt = initparsopt, 
  exteq = exteq,
  soc = soc, 
  cond = cond,
  btorph = btorph
)
```

The ML parameter estimates are:

```{r}
print(r)


